media-proxy/src/librabric_ep.c

```diff
@@ -0,0 +1,314 @@
+#include <assert.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <poll.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sched.h>
+#include <sys/wait.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+
+#include <rdma/fi_cm.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_rma.h>
+#include <rdma/fi_tagged.h>
+#include <rdma/fi_atomic.h>
+#include <rdma/fi_collective.h>
+
+#include "libfabric_ep.h"
+#include "libfabric_cq.h"
+#include "libfabric_mr.h"
+
+#define RDMA_CTRL_MSG_LEN 1024
+
+
+static int enable_ep(ep_ctx_t* ep_ctx)
+{
+	uint64_t flags;
+	int ret;
+
+	RDMA_EP_BIND(ep_ctx->ep, ep_ctx->av, 0);
+
+	flags = FI_TRANSMIT;
+	RDMA_EP_BIND(ep_ctx->ep, ep_ctx->txcq, flags);
+
+	flags = FI_RECV;
+	RDMA_EP_BIND(ep_ctx->ep, ep_ctx->rxcq, flags);
+
+	ret = rdma_get_cq_fd(ep_ctx->txcq, &ep_ctx->tx_fd, ep_ctx->rdma_ctx->comp_method);
+	if (ret)
+		return ret;
+
+	ret = rdma_get_cq_fd(ep_ctx->rxcq, &ep_ctx->rx_fd, ep_ctx->rdma_ctx->comp_method);
+	if (ret)
+		return ret;
+
+	ret = fi_enable(ep_ctx->ep);
+	if (ret) {
+		RDMA_PRINTERR("fi_enable", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int ep_av_insert(libfabric_ctx* rdma_ctx, struct fid_av *av, void *addr, size_t count, fi_addr_t *fi_addr,
+		uint64_t flags, void *context)
+{
+	int ret;
+
+	ret = fi_av_insert(av, addr, count, fi_addr, flags, context);
+	if (ret < 0) {
+		RDMA_PRINTERR("fi_av_insert", ret);
+		return ret;
+	} else if (ret != count) {
+		RDMA_ERR("fi_av_insert: number of addresses inserted = %d;"
+			       " number of addresses given = %zd\n", ret, count);
+		return -EXIT_FAILURE;
+	}
+
+	return 0;
+}
+
+
+int ep_init_av_addr(ep_ctx_t* ep_ctx, libfabric_ctx* rdma_ctx, struct fi_info* fi)
+{
+    size_t addrlen;
+    int ret;
+
+    if (fi->dest_addr) {
+        ret = ep_av_insert(rdma_ctx, ep_ctx->av, fi->dest_addr, 1, &ep_ctx->dest_av_entry, 0, NULL);
+        if (ret)
+            return ret;
+
+        addrlen = RDMA_CTRL_MSG_LEN;
+        ret = fi_getname(&ep_ctx->ep->fid, (char *) ep_ctx->data_buf,
+                 &addrlen);
+        if (ret) {
+            RDMA_PRINTERR("fi_getname", ret);
+            return ret;
+        }
+        struct sockaddr_in* insert_addr_p = (struct sockaddr_in*)ep_ctx->data_buf;
+        memset(insert_addr_p->sin_zero, 0, sizeof(insert_addr_p->sin_zero));
+
+        ret = ep_send_buf(ep_ctx, ep_ctx->data_buf, addrlen);
+        if (ret)
+            return ret;
+
+    } else {
+        ret = ep_recv_buf(ep_ctx, ep_ctx->data_buf, RDMA_CTRL_MSG_LEN);
```

Comment:
No we can't pass RDMA_CTRL_MSG_LEN directly as fi_getname uses addrlen as an input and output. On input it is capacity of the provided buffer, on output it is actual size of the returned address.

Now when I look at it I see that using arbitrary size is not needed and instead of RDMA_CTRL_MSG_LEN I will set the addrlen to ep_ctx->data_buf_size